### PR TITLE
QUIC: Immediate-Mode Polling API

### PR DESCRIPTION
**Pull request by hlandau**
Fixes https://github.com/openssl/project/issues/299.
Built on #24257.
Review #24257 first.

```
d8385dd9e4 QUIC REACTOR: Add support for external registration of blocking operations
8ec02de2d5 QUIC REACTOR: Add utility for tracking recursive blocking operations
56d6ebf7b8 QUIC APL: Add support for registering blocking operations to support polling code
30af4c907d RIO: Add poll builder to support immediate-mode polling API
2375d0b65e RIO: Amend SSL_poll to support blocking on QUIC objects
1b81bcf33a QUIC POLLING: Add support for polling listeners
3e67ccf9aa RIO: Amend SSL_poll code to correctly register blocking operations for inter-thread notification
7d9072be79 QUIC APL: Fix various functions when called on listeners
fa73f213f5 QUIC MULTISTREAM TEST: Remove test that blocking SSL_poll doesn't work
70195836be QUIC RADIX: Test new SSL_poll functionality
298de370d0 QUIC RADIX: Test listener polling support
44cb8d9069 QUIC RADIX: Use enhanced blocking support when testing
601ce8c6ab QUIC: Glossary updates
6af342547d QUIC: Update SSL_poll documentation
```

### CODE REVIEW

File: ssl/quic/quic_impl.c [hunk 1]
```diff
@@ -4163,6 +4231,60 @@ SSL *ossl_quic_new_listener(SSL_CTX *ctx, uint64_t flags)
...
+    port_args.is_multi_conn     = 1;
+    ql->port = ossl_quic_engine_create_port(ctx.qd->engine, &port_args);
+    if (ql->port == NULL) {
+        QUIC_RAISE_NON_NORMAL_ERROR(NULL, ERR_R_INTERNAL_ERROR, NULL);
+        goto err;
+    }
+
+    ql->domain  = ctx.qd;
+    ql->engine  = ctx.qd->engine;
+    ql->mutex   = ctx.qd->mutex;
```
Comment (nhorman) [inline ssl/quic/quic_impl.c:4304]: the definition of ql->mutex is gated on OPENSSL_THREADS.  A build configured with no-threads will break here as ->mutex will be undefined

File: ssl/rio/rio_notifier.c [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+    fd = BIO_socket(domain, socktype, protocol, 0);
+    if (fd < 0)
+        return -1;
+
+    /*
+     * Make socket close-on-exec unless this was already done above at socket
+     * creation time.
+     */
+    if (!set_cloexec(fd)) {
```
Comment (nhorman) [inline ssl/rio/rio_notifier.c:57]: I'm not sure I understand this comment.  The socket is created by this function, so this is socket creation time, no?

File: ssl/rio/rio_notifier.c [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+    /*
+     * Sanity check - ensure someone else didn't connect to our listener during
+     * the brief window of possibility above.
+     */
+    if (accept_sa.sin_family != AF_INET || accept_sa.sin_port != sa.sin_port)
+        goto err;
+
+    /* Make both sides of the connection non-blocking. */
+    if (!BIO_socket_nbio(rfd, 1))
```
Comment (nhorman) [inline ssl/rio/rio_notifier.c:145]: If we're doing non-blocking io on a tcp socket here, wouldn't it be simpler to just use a dgram pair?

File: ssl/quic/quic_reactor.c [hunk 4]
```diff
@@ -397,12 +464,77 @@ int ossl_quic_reactor_block_until_pred(QUIC_REACTOR *rtor,
...
+         *   Second, the thread which happened to be the one which decremented
+         *   cur_blocking_waiters to 0 unsignals the notifier and is then
+         *   responsible for broadcasting to a CV to indicate to the other
+         *   threads that the synchronised wakeup has been cmpleted. Other
+         *   threads wait for this CV to be signalled.
+         *
+         */
+        if (rtor->have_notifier && rtor->signalled_notifier) {
+            if (rtor->cur_blocking_waiters == 0) {
+                ossl_rio_notifier_unsignal(&rtor->notifier);
```
Comment (nhorman) [inline ssl/quic/quic_reactor.c]: I can't put my finger on the exact reason why, but having to do un-signaling here feels wrong.  need to think about this some more

**Review by nhorman — CHANGES_REQUESTED** [review]

File: ssl/rio/rio_notifier.c (hunk 2 above)
Comment (levitte) [reply ssl/rio/rio_notifier.c:57]: The socket is created with the `BIO_socket()` call on new line 49.

File: ssl/rio/rio_notifier.c (hunk 3 above)
Comment (levitte) [reply ssl/rio/rio_notifier.c:145]: non-blocking TCP has other semantics than dgram (UDP), yeah?

File: ssl/rio/rio_notifier.c [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+    int fds[2], domain = AF_INET, type = SOCK_STREAM;
+
+# if defined(SOCK_CLOEXEC)
+    type |= SOCK_CLOEXEC;
+# endif
+# if defined(SOCK_NONBLOCK)
+    type |= SOCK_NONBLOCK;
+# endif
+
+#if defined(OPENSSL_SYS_UNIX) && defined(AF_UNIX)
```
[the remark this answers is not in the document]
Comment (levitte) [reply ssl/rio/rio_notifier.c:178]: You need to look at `include/openssl/e_os2.h` as well.  It starts with defining `OPENSSL_SYS_UNIX`, and then undefines it for systems that categorically doesn't fall into the Unix family, even by a little.

As you can see yourself, that header file doesn't undef `OPENSSL_SYS_UNIX` for MacOSX.

File: ssl/rio/poll_builder.c [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        openssl_fdset(fd, &rpb->wfd);
+
+    openssl_fdset(fd, &rpb->efd);
+    if (fd > rpb->hwm_fd)
+        rpb->hwm_fd = fd;
+
+    return 1;
+
+#elif RIO_POLL_METHOD == RIO_POLL_METHOD_POLL
+    size_t num_loop;
```
Comment (mattcaswell) [inline ssl/rio/poll_builder.c:86]: Mixed declarations and code here

File: test/radix/quic_tests.c [hunk 7]
```diff
@@ -119,6 +121,11 @@ DEF_FUNC(ssl_poll_check)
         timeout.tv_sec  = 10;
         timeout.tv_usec = 0;
         break;
+    case 4: /* Listener test */
+        expected_result_count       = 1;
+        items[5].events            |= SSL_POLL_EVENT_IC;
```
Comment (mattcaswell) [inline test/radix/quic_tests.c:126]: Why `|=` here, since we haven't previously set this to anything?

File: doc/man3/SSL_poll.pod [hunk 8]
```diff
@@ -278,29 +319,19 @@ unidirectional stream to be locally created.
...
-NULL, will fail. This does not allow waiting, but does allow multiple QUIC SSL
-objects to be queried for their readiness state in a single call.
-
-Future releases will remove this limitation and support blocking SSL_poll().
-
-=item
-
 Only B<BIO_POLL_DESCRIPTOR> structures with type
 B<BIO_POLL_DESCRIPTOR_TYPE_SSL>, referencing QUIC connection SSL objects or QUIC
 stream SSL objects, are supported.
```
Comment (mattcaswell) [inline doc/man3/SSL_poll.pod:330]: Shouldn't this also reference QUIC Listener SSL Objects?

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Finished my review. A couple more comments.

**Pull request closed without merging**